File: go-fullstack/backend/handlers/stats.go
package handlers

import (
	"context"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/locus/go-fullstack-backend/cache"
	"github.com/locus/go-fullstack-backend/database"
)

// StatsResponse represents the statistics response
type StatsResponse struct {
	PageViews  int `json:"pageViews"`
	TotalTodos int `json:"totalTodos"`
}

// GetStats handles GET /api/stats
func GetStats(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), defaultTimeout)
	defer cancel()

	// Increment page view counter in Redis
	cache.Client.Incr(ctx, "stats:page-views")
	pageViews, err := cache.Client.Get(ctx, "stats:page-views").Int()
	if err != nil {
		pageViews = 0
	}

	// Get total count
	var totalTodos int
	err = database.DB.QueryRow(ctx, `SELECT COUNT(*) FROM todos`).Scan(&totalTodos)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch statistics"})
		return
	}

	c.JSON(http.StatusOK, StatsResponse{
		PageViews:  pageViews,
		TotalTodos: totalTodos,
	})
}

File: rust-backend/backend/src/handlers/stats.rs
use axum::{extract::State, Json};
use serde::Serialize;

use crate::{db, error::Result, AppState};

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StatsResponse {
    pub page_views: i64,
    pub total_todos: i64,
}

pub async fn get_stats(State(state): State<AppState>) -> Result<Json<StatsResponse>> {
    // Increment page view counter in Redis
    let _ = state.cache.incr("stats:page-views").await;
    let page_views = state
        .cache
        .get("stats:page-views")
        .await
        .ok()
        .flatten()
        .and_then(|v| v.parse::<i64>().ok())
        .unwrap_or(0);

    let total_todos = db::get_total_todo_count(&state.db).await?;

    Ok(Json(StatsResponse {
        page_views,
        total_todos,
    }))
}
